flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/stream/StreamingJoinOperator.java

```diff
@@ -0,0 +1,476 @@
+					}
+				}
+			}
+		}
+	}
+
+	// -------------------------------------------------------------------------------------
+
+	private void output(BaseRow inputRow, BaseRow otherRow, boolean inputIsLeft) {
+		if (inputIsLeft) {
+			outRow.replace(inputRow, otherRow);
+		} else {
+			outRow.replace(otherRow, inputRow);
+		}
+		collector.collect(outRow);
+	}
+
+	private void outputNullPadding(BaseRow row, boolean isLeft) {
+		if (isLeft) {
+			outRow.replace(row, rightNullRow);
+		} else {
+			outRow.replace(leftNullRow, row);
+		}
+		collector.collect(outRow);
+	}
+
+	// ----------------------------------------------------------------------------------------
+	// Utility Classes
+	// ----------------------------------------------------------------------------------------
+
+	private class JoinConditionWithNullFilters implements JoinCondition {
+
+		final JoinCondition backingJoinCondition;
+
+		private JoinConditionWithNullFilters(JoinCondition backingJoinCondition) {
+			this.backingJoinCondition = backingJoinCondition;
+		}
+
+		@Override
+		public boolean apply(BaseRow left, BaseRow right) {
+			if (!nullSafe) { // is not null safe, return false if any null exists
+				// key is always BinaryRow
+				BinaryRow joinKey = (BinaryRow) getCurrentKey();
+				if (filterAllNulls ? joinKey.anyNull() : joinKey.anyNull(nullFilterKeys)) {
+					// find null present, return false directly
+					return false;
+				}
+			}
+			// test condition
+			return backingJoinCondition.apply(left, right);
+		}
+	}
+
+	private static final class AssociatedRecords {
+		private final List<OuterRecord> records;
+
+		private AssociatedRecords(List<OuterRecord> records) {
+			checkNotNull(records);
+			this.records = records;
+		}
+
+		public boolean isEmpty() {
+			return records.isEmpty();
+		}
+
+		public int size() {
+			return records.size();
+		}
+
+		public Iterable<BaseRow> getRecords() {
+			return new RecordsIterable(records);
+		}
+
+		public Iterable<OuterRecord> getOuterRecords() {
+			return records;
+		}
+
+		public static AssociatedRecords of(
+			BaseRow input,
+			boolean inputIsLeft,
+			JoinRecordStateView otherSideStateView,
+			JoinCondition condition) throws Exception {
+
+			List<OuterRecord> associations = new ArrayList<>();
+			if (otherSideStateView instanceof OuterJoinRecordStateView) {
+				OuterJoinRecordStateView outerStateView = (OuterJoinRecordStateView) otherSideStateView;
+				Iterable<Tuple2<BaseRow, Integer>> records = outerStateView.getRecordsAndNumOfAssociations();
+				for (Tuple2<BaseRow, Integer> record : records) {
+					boolean matched = inputIsLeft ? condition.apply(input, record.f0) : condition.apply(record.f0, input);
+					if (matched) {
+						associations.add(new OuterRecord(record.f0, record.f1));
+					}
+				}
+			} else {
+				Iterable<BaseRow> records = otherSideStateView.getRecords();
+				for (BaseRow record : records) {
```

Comment:
NIT: Maybe you can try a lazy transformation? Let `AssociatedRecords` wrap `Iterable<BaseRow>`?

Reply:
I think we can't. Because we need to know whether there is any associated records, i.e. `AssociatedRecords.isEmpty`. So we have to eagerly iterate the records.